Tsconfig refactor

MaxChen102 opened the pull request:
## Proposed Changes

Refactoring code to allow strict null checks

...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| 🎨 _User interface change_ (change to user interface; provide screenshots)              |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |     X     |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [ ] I have performed a self-review of my changes.
    - Check that all changed files included in this pull request are intentional changes.
    - Check that all changes are relevant to the purpose of this pull request, as described above.
- [ ] I have added tests for my changes, if applicable.
    - This is **required** for all bug fixes and new features.
- [ ] I have updated the project documentation, if applicable.
    - This is **required** for new features.
- [ ] If this is my first contribution, I have added myself to the list of contributors.
- [ ] I have updated the project Changelog (this is required for all changes).

After opening your pull request:

- [ ] I have verified that the CI tests have passed.
- [ ] I have reviewed the test coverage changes reported by Coveralls.
- [ ] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

david-yz-liu commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
 
+- Added Typescript support for "strictNullChecks" and refactored code accordingly
```
Entries should go under one of the subsections of the changelog. Your work for his PR is an internal change.

david-yz-liu commented on memory-viz/src/user_functions.ts:
```diff
@@ -48,6 +48,9 @@ function draw(
 
     if (typeof objects === "string") {
+        if (!fs) {
+            throw new Error("fs module not available in this environment.");
```
Let's make this error message more informative: "Could not load file {objects} in this environment."

david-yz-liu commented on memory-viz/src/tests/plugins/pretty_svg.ts:
```diff
@@ -11,7 +11,10 @@ export const serialize: NewPlugin["serialize"] = (val: HTMLElement) => {
 export const test: NewPlugin["test"] = (val: any) => {
     const root = parse(val);
-    return root.firstChild.rawTagName === "svg";
+    if (root.firstChild !== undefined) {
```
It's a bit cleaner here to just use the `?.` operator to acccess `rawTagName`, since `undefined === "svg"` will evaluate to `false`.

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -88,7 +88,9 @@ export class MemoryModel {
             options.height ? options.height.toString() : "800"
         );
-        this.roughjs_config = options.roughjs_config;
+        this.roughjs_config = options.roughjs_config ?? {
```
There should not be a default seed value. Use just `{}` as the default `this.roughjs_config` value.

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -122,6 +124,9 @@ export class MemoryModel {
             console.log(xml);
         } else {
+            if (!fs) {
+                throw new Error("fs module not available in this environment.");
```
See my comment about the fs error message below.

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -176,7 +185,11 @@ export class MemoryModel {
         style: Style
     ): Rect {
         if (collections.includes(type)) {
-            if (type === "dict" && typeof value === "object") {
```
I would actually like to allow `null` values to be passed to `drawDict`. Please modify the type signature for that function, and if `null` is passed in, treat it like an empty object `{}` (i.e., no keys/values are rendered).

MaxChen102 replied:
<img width="515" alt="image" src="https://github.com/user-attachments/assets/d4f245a6-319b-4a70-a838-8c91a3f9a973" />
This is what the frame looks like when {} is passed in for value, should I keep it like this when null is passed in? or should the frame be a little longer?

david-yz-liu replied:
The behaviour should be identical when `null` vs. `{}` are passed in

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -927,27 +940,27 @@ export class MemoryModel {
-                    obj.id,
+                    obj.x!,
+                    obj.y!,
+                    obj.name!,
```
Allow this to be undefined/null

david-yz-liu replied:
All changes are good, except I think you missed this comment. the `name` parameter for `drawClass` should allow for `undefined`/`null`

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -927,27 +940,27 @@ export class MemoryModel {
+                    obj.x!,
+                    obj.y!,
+                    obj.name!,
+                    obj.id!,
```
Allow this to be undefined/null

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -927,27 +940,27 @@ export class MemoryModel {
+                    obj.x!,
+                    obj.y!,
+                    obj.type!,
+                    obj.id!,
```
Allow this to be undefined/null

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -927,27 +940,27 @@ export class MemoryModel {
+                    obj.id!,
                     obj.value,
-                    obj.show_indexes,
+                    obj.show_indexes!,
```
Allow this to be undefined/null (default value `false`)

david-yz-liu commented on memory-viz/src/automate.ts:
```diff
@@ -387,6 +423,14 @@ function compareByRightness(a: DrawnEntity, b: DrawnEntity): number {
+        b.y === undefined ||
+        b.height === undefined
+    ) {
+        throw new Error("Both objects must have 'x' and 'width' property.");
```
This error message is incorrect

david-yz-liu commented on memory-viz/src/automate.ts:
```diff
@@ -41,9 +41,9 @@ function drawAutomated(
     }
 
-    min_width += requiredWidth + 2 * configuration.padding + 1;
+    min_width += requiredWidth + 2 * (configuration.padding ?? 25) + 1;
```
Define a constant in this file for `DEFAULT_PADDING`.

david-yz-liu reviewed (commented):
@MaxChen102 really nice work! I left a few inline comments, but overall this is a great improvement to the codebase.

MaxChen102 commented:
Hi @david-yz-liu , I've updated the code based on your feedback. I've updated drawObject and drawClass to allow for undefined arguments but I am unsure why we need to allow this. Since the docstring of the function drawAll (which then calls drawObject and drawClass) states that every object must include those attributes.

david-yz-liu commented on memory-viz/src/automate.ts:
```diff
@@ -183,6 +187,16 @@ function drawAutomatedOtherItems(
+            item.type === ".blank" &&
+            (item.width === undefined || item.height === undefined)
+        ) {
+            console.log(
```
Use `console.warn` here

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -167,39 +176,41 @@ export class MemoryModel {
+        x: number | undefined = 0,
+        y: number | undefined = 0,
+        type: string | undefined,
+        id: number | undefined = 0,
```
It looks like you misunderstood my previous comment. My comments always refer to one specific line (unless I specify otherwise, or select a range of lines).

The only parameters I wanted you to make optional were `id` and `show_indexes`. For `id`, do not specify a default value; the code itself should handle when `id === undefined`.

david-yz-liu reviewed (commented):
@MaxChen102 I left a few more comments. Note that I clarified exactly which parameters you should make optional; if you believe the documentation or type annotations for `drawAll` needs to be updated, you can do that as well, but make sure you understand exactly what change I'm asking you to make before doing so.

In addition to these comments, please also update this branch with the latest from master, and make sure your changes are compatible with the ones made in #146.

MaxChen102 commented:
Hi @david-yz-liu , thanks for pointing my mistakes out and I have fixed them, please let me know what you think.

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -966,18 +981,17 @@ export class MemoryModel {
         configuration: Partial<DisplaySettings>,
         snapshotObjects: DrawnEntity[]
     ): Size {
-        // Dynamically determining the width of the canvas, in case one has not been provided.
```
Keep this comment

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -986,18 +1000,19 @@ export class MemoryModel {
             size.width = rightmost_edge + 100;
         }
 
-        // Dynamically determining the height of the canvas, in case one has not been provided.
```
Keep this comment
